[BUG]: void value not ignored as it ought to be

### Required prerequisites

- [X] Make sure you've read the [documentation](https://pybind11.readthedocs.io). Your issue may be addressed there.
- [X] Search the [issue tracker](https://github.com/pybind/pybind11/issues) and [Discussions](https:/pybind/pybind11/discussions) to verify that this hasn't already been reported. +1 or comment there if it has.
- [ ] Consider asking first in the [Gitter chat room](https://gitter.im/pybind/Lobby) or in a [Discussion](https:/pybind/pybind11/discussions/new).

### Problem description

the bind code like this :
```cpp
#include <pybind11/pybind11.h>
#include <iostream>
namespace py = pybind11;
using std::cout;
using std::endl;
class Animal {
public:
    virtual ~Animal() { }
    virtual void* go() = 0;
};

class PyAnimal : public Animal {
public:
    /* Inherit the constructors */
    using Animal::Animal;
    /* Trampoline (need one for each virtual function) */
    void* go() override {
        PYBIND11_OVERRIDE_PURE(
            void*, /* Return type */
            Animal,      /* Parent class */
            go,          /* Name of function in C++ (must match Python name) */
        );
    }
};

PYBIND11_MODULE(example, m) {
    py::class_<Animal, PyAnimal /* <--- trampoline*/>(m, "Animal")
        .def(py::init<>())
        .def("go", &Animal::go);
}
```

use pybind11 2.10.0,get error below:

```
example.cpp: In member function ‘virtual void* PyAnimal::go(int)’:
example.cpp:21:9: error: void value not ignored as it ought to be
   21 |         PYBIND11_OVERRIDE_PURE(
      |         ^~~~~~~~~~~~~~~~~~~~~~
error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
```

but use pybind 2.9.2 is ok ，how to deal with it ？？

### Reproducible example code

_No response_

I get

```python
tmp.cpp:18:9: error: cannot initialize return object of type 'void *' with an rvalue of type 'enable_if_t<std::is_same<void, intrinsic_t<void *>>::value, void>' (aka 'void')
        PYBIND11_OVERRIDE_PURE(
        ^~~~~~~~~~~~~~~~~~~~~~~
/Users/henryschreiner/git/software/pybind11/include/pybind11/pybind11.h:2826:5: note: expanded from macro 'PYBIND11_OVERRIDE_PURE'
    PYBIND11_OVERRIDE_PURE_NAME(                                                                  \
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/henryschreiner/git/software/pybind11/include/pybind11/pybind11.h:2788:9: note: expanded from macro 'PYBIND11_OVERRIDE_PURE_NAME'
        PYBIND11_OVERRIDE_IMPL(PYBIND11_TYPE(ret_type), PYBIND11_TYPE(cname), name, __VA_ARGS__); \
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/henryschreiner/git/software/pybind11/include/pybind11/pybind11.h:2754:20: note: expanded from macro 'PYBIND11_OVERRIDE_IMPL'
            return pybind11::detail::cast_safe<ret_type>(std::move(o));                           \
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

Debug mechanism:

```console
$ cmd_to_run=$(python3 -m pybind11 --extension=example.cpp)
$ git bisect start
$ git bisect good v2.9.2
$ git bisect bad v2.10.0
$ git bisect run $cmd_to_run
running  'clang++' '-I/usr/local/opt/python@3.10/Frameworks/Python.framework/Versions/3.10/include/python3.10' '-I/usr/local/opt/python@3.10/Frameworks/Python.framework/Versions/3.10/include/python3.10' '-I/Users/henryschreiner/git/software/pybind11/include' '-Wno-unused-result' '-Wsign-compare' '-Wunreachable-code' '-fno-common' '-dynamic' '-DNDEBUG' '-g' '-fwrapv' '-O3' '-Wall' '-isysroot' '/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk' '-std=c++11' '-isysroot' '/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk' '-undefined' 'dynamic_lookup' 'example.cpp' '-o' 'example.cpython-310-darwin.so'
Bisecting: 60 revisions left to test after this (roughly 6 steps)
[03252067db057777db75c2e5bcf0e69013b919ec] chore(deps): bump actions/upload-artifact from 2 to 3 (#3899)
running  'clang++' '-I/usr/local/opt/python@3.10/Frameworks/Python.framework/Versions/3.10/include/python3.10' '-I/usr/local/opt/python@3.10/Frameworks/Python.framework/Versions/3.10/include/python3.10' '-I/Users/henryschreiner/git/software/pybind11/include' '-Wno-unused-result' '-Wsign-compare' '-Wunreachable-code' '-fno-common' '-dynamic' '-DNDEBUG' '-g' '-fwrapv' '-O3' '-Wall' '-isysroot' '/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk' '-std=c++11' '-isysroot' '/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk' '-undefined' 'dynamic_lookup' 'example.cpp' '-o' 'example.cpython-310-darwin.so'
example.cpp:18:9: error: cannot initialize return object of type 'void *' with an rvalue of type 'enable_if_t<std::is_same<void, intrinsic_t<void *>>::value, void>' (aka 'void')
        PYBIND11_OVERRIDE_PURE(
        ^~~~~~~~~~~~~~~~~~~~~~~
/Users/henryschreiner/git/software/pybind11/include/pybind11/pybind11.h:2808:5: note: expanded from macro 'PYBIND11_OVERRIDE_PURE'
    PYBIND11_OVERRIDE_PURE_NAME(                                                                  \
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/henryschreiner/git/software/pybind11/include/pybind11/pybind11.h:2770:9: note: expanded from macro 'PYBIND11_OVERRIDE_PURE_NAME'
        PYBIND11_OVERRIDE_IMPL(PYBIND11_TYPE(ret_type), PYBIND11_TYPE(cname), name, __VA_ARGS__); \
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/henryschreiner/git/software/pybind11/include/pybind11/pybind11.h:2736:20: note: expanded from macro 'PYBIND11_OVERRIDE_IMPL'
            return pybind11::detail::cast_safe<ret_type>(std::move(o));                           \
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
Bisecting: 30 revisions left to test after this (roughly 5 steps)
[b22ee64c7309eb7344031b5e5385e6907d84c7eb] Add type_caster<std::monostate> (#3818)
running  'clang++' '-I/usr/local/opt/python@3.10/Frameworks/Python.framework/Versions/3.10/include/python3.10' '-I/usr/local/opt/python@3.10/Frameworks/Python.framework/Versions/3.10/include/python3.10' '-I/Users/henryschreiner/git/software/pybind11/include' '-Wno-unused-result' '-Wsign-compare' '-Wunreachable-code' '-fno-common' '-dynamic' '-DNDEBUG' '-g' '-fwrapv' '-O3' '-Wall' '-isysroot' '/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk' '-std=c++11' '-isysroot' '/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk' '-undefined' 'dynamic_lookup' 'example.cpp' '-o' 'example.cpython-310-darwin.so'
Bisecting: 15 revisions left to test after this (roughly 4 steps)
[9969f3b5b5ccad5524a9dab237b551c0ea6359f4] ci: drop win2016 (#3854)
running  'clang++' '-I/usr/local/opt/python@3.10/Frameworks/Python.framework/Versions/3.10/include/python3.10' '-I/usr/local/opt/python@3.10/Frameworks/Python.framework/Versions/3.10/include/python3.10' '-I/Users/henryschreiner/git/software/pybind11/include' '-Wno-unused-result' '-Wsign-compare' '-Wunreachable-code' '-fno-common' '-dynamic' '-DNDEBUG' '-g' '-fwrapv' '-O3' '-Wall' '-isysroot' '/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk' '-std=c++11' '-isysroot' '/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk' '-undefined' 'dynamic_lookup' 'example.cpp' '-o' 'example.cpython-310-darwin.so'
example.cpp:18:9: error: cannot initialize return object of type 'void *' with an rvalue of type 'enable_if_t<std::is_same<void, intrinsic_t<void *>>::value, void>' (aka 'void')
        PYBIND11_OVERRIDE_PURE(
        ^~~~~~~~~~~~~~~~~~~~~~~
/Users/henryschreiner/git/software/pybind11/include/pybind11/pybind11.h:2808:5: note: expanded from macro 'PYBIND11_OVERRIDE_PURE'
    PYBIND11_OVERRIDE_PURE_NAME(                                                                  \
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/henryschreiner/git/software/pybind11/include/pybind11/pybind11.h:2770:9: note: expanded from macro 'PYBIND11_OVERRIDE_PURE_NAME'
        PYBIND11_OVERRIDE_IMPL(PYBIND11_TYPE(ret_type), PYBIND11_TYPE(cname), name, __VA_ARGS__); \
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/henryschreiner/git/software/pybind11/include/pybind11/pybind11.h:2736:20: note: expanded from macro 'PYBIND11_OVERRIDE_IMPL'
            return pybind11::detail::cast_safe<ret_type>(std::move(o));                           \
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
Bisecting: 7 revisions left to test after this (roughly 3 steps)
[65ec5de52fecc974157a0c28490fef61204c39a1] chore: bump changelog for 2.9.2 (#3834)
running  'clang++' '-I/usr/local/opt/python@3.10/Frameworks/Python.framework/Versions/3.10/include/python3.10' '-I/usr/local/opt/python@3.10/Frameworks/Python.framework/Versions/3.10/include/python3.10' '-I/Users/henryschreiner/git/software/pybind11/include' '-Wno-unused-result' '-Wsign-compare' '-Wunreachable-code' '-fno-common' '-dynamic' '-DNDEBUG' '-g' '-fwrapv' '-O3' '-Wall' '-isysroot' '/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk' '-std=c++11' '-isysroot' '/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk' '-undefined' 'dynamic_lookup' 'example.cpp' '-o' 'example.cpython-310-darwin.so'
Bisecting: 3 revisions left to test after this (roughly 2 steps)
[ab59f45d2ee05b1df299021b62ab1f59a768e285] Prefer make_caster<T> to type_caster<T> (#3859)
running  'clang++' '-I/usr/local/opt/python@3.10/Frameworks/Python.framework/Versions/3.10/include/python3.10' '-I/usr/local/opt/python@3.10/Frameworks/Python.framework/Versions/3.10/include/python3.10' '-I/Users/henryschreiner/git/software/pybind11/include' '-Wno-unused-result' '-Wsign-compare' '-Wunreachable-code' '-fno-common' '-dynamic' '-DNDEBUG' '-g' '-fwrapv' '-O3' '-Wall' '-isysroot' '/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk' '-std=c++11' '-isysroot' '/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk' '-undefined' 'dynamic_lookup' 'example.cpp' '-o' 'example.cpython-310-darwin.so'
Bisecting: 1 revision left to test after this (roughly 1 step)
[e3aa215b020886d648add951186052c619c3cf9d] Add perfect forwarding to make_iterator calls (#3860)
running  'clang++' '-I/usr/local/opt/python@3.10/Frameworks/Python.framework/Versions/3.10/include/python3.10' '-I/usr/local/opt/python@3.10/Frameworks/Python.framework/Versions/3.10/include/python3.10' '-I/Users/henryschreiner/git/software/pybind11/include' '-Wno-unused-result' '-Wsign-compare' '-Wunreachable-code' '-fno-common' '-dynamic' '-DNDEBUG' '-g' '-fwrapv' '-O3' '-Wall' '-isysroot' '/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk' '-std=c++11' '-isysroot' '/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk' '-undefined' 'dynamic_lookup' 'example.cpp' '-o' 'example.cpython-310-darwin.so'
Bisecting: 0 revisions left to test after this (roughly 0 steps)
[088ad4f2987d3da4b1ffab376b2b4812fd7634ef] Cleanup cast_safe<void> specialization (#3861)
running  'clang++' '-I/usr/local/opt/python@3.10/Frameworks/Python.framework/Versions/3.10/include/python3.10' '-I/usr/local/opt/python@3.10/Frameworks/Python.framework/Versions/3.10/include/python3.10' '-I/Users/henryschreiner/git/software/pybind11/include' '-Wno-unused-result' '-Wsign-compare' '-Wunreachable-code' '-fno-common' '-dynamic' '-DNDEBUG' '-g' '-fwrapv' '-O3' '-Wall' '-isysroot' '/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk' '-std=c++11' '-isysroot' '/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk' '-undefined' 'dynamic_lookup' 'example.cpp' '-o' 'example.cpython-310-darwin.so'
example.cpp:18:9: error: cannot initialize return object of type 'void *' with an rvalue of type 'enable_if_t<std::is_same<void, intrinsic_t<void *>>::value, void>' (aka 'void')
        PYBIND11_OVERRIDE_PURE(
        ^~~~~~~~~~~~~~~~~~~~~~~
/Users/henryschreiner/git/software/pybind11/include/pybind11/pybind11.h:2808:5: note: expanded from macro 'PYBIND11_OVERRIDE_PURE'
    PYBIND11_OVERRIDE_PURE_NAME(                                                                  \
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/henryschreiner/git/software/pybind11/include/pybind11/pybind11.h:2770:9: note: expanded from macro 'PYBIND11_OVERRIDE_PURE_NAME'
        PYBIND11_OVERRIDE_IMPL(PYBIND11_TYPE(ret_type), PYBIND11_TYPE(cname), name, __VA_ARGS__); \
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/henryschreiner/git/software/pybind11/include/pybind11/pybind11.h:2736:20: note: expanded from macro 'PYBIND11_OVERRIDE_IMPL'
            return pybind11::detail::cast_safe<ret_type>(std::move(o));                           \
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
struct override_unused {};
1 error generated.
088ad4f2987d3da4b1ffab376b2b4812fd7634ef is the first bad commit
commit 088ad4f2987d3da4b1ffab376b2b4812fd7634ef
Author: Laramie Leavitt <laramiel@users.noreply.github.com>
Date:   Mon Apr 11 11:57:05 2022 -0700

    Cleanup cast_safe<void> specialization (#3861)

    * Cleanup cast_safe<void> specialization

    Replace explicit specialization of cast_safe<void> with SFINAE.
    It's better for SFINAE cases to cover all type-sets rather than mixing SFINAE and explicit specialization.

    Extracted from #3674

    * [pre-commit.ci] auto fixes from pre-commit.com hooks

    for more information, see https://pre-commit.ci

    * Update cast.h

    Use detail::none_of<> as suggested
struct override_unused {};

  struct override_unused {};
    * [pre-commit.ci] auto fixes from pre-commit.com hooks

    for more information, see https://pre-commit.ci

    * Update cast.h

    Reorder:
    If TEMP_REF
    If VOID
    if (!VOID && !TEMP_REF)

    Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

 include/pybind11/cast.h | 15 +++++++++------
 1 file changed, 9 insertions(+), 6 deletions(-)
bisect found first bad commit⏎
  // Trampoline use: for reference/pointer types to value-converted values, we do a value cast, then
```

--- a/include/pybind11/cast.h
+++ b/include/pybind11/cast.h
@@ -1179,11 +1179,9 @@ enable_if_t<cast_is_temporary_value_reference<T>::value, T> cast_safe(object &&)
     pybind11_fail("Internal error: cast_safe fallback invoked");
 }
 template <typename T>
-enable_if_t<std::is_same<void, intrinsic_t<T>>::value, void> cast_safe(object &&) {}
+enable_if_t<std::is_same<void, T>::value, void> cast_safe(object &&) {}
 template <typename T>
-enable_if_t<detail::none_of<cast_is_temporary_value_reference<T>,
-                            std::is_same<void, intrinsic_t<T>>>::value,
-            T>
+enable_if_t<detail::none_of<cast_is_temporary_value_reference<T>, std::is_same<void, T>>::value, T>
 cast_safe(object &&o) {
     return pybind11::cast<T>(std::move(o));
 }

--- a/include/pybind11/cast.h
+++ b/include/pybind11/cast.h
@@ -1179,9 +1179,9 @@ enable_if_t<cast_is_temporary_value_reference<T>::value, T> cast_safe(object &&)
     pybind11_fail("Internal error: cast_safe fallback invoked");
 }
 template <typename T>
-enable_if_t<std::is_same<void, T>::value, void> cast_safe(object &&) {}
+enable_if_t<std::is_void<T>::value, void> cast_safe(object &&) {}
 template <typename T>
-enable_if_t<detail::none_of<cast_is_temporary_value_reference<T>, std::is_same<void, T>>::value, T>
+enable_if_t<detail::none_of<cast_is_temporary_value_reference<T>, std::is_void<T>>::value, T>
 cast_safe(object &&o) {
     return pybind11::cast<T>(std::move(o));
 }

--- a/tests/test_class.cpp
+++ b/tests/test_class.cpp
@@ -385,6 +385,7 @@ TEST_SUBMODULE(class_, m) {
     protected:
         virtual int foo() const { return value; }
         virtual void *void_foo() { return (void *) &value; }
+        virtual void *get_self() { return (void *) this; }
 
     private:
         int value = 42;
@@ -394,6 +395,7 @@ TEST_SUBMODULE(class_, m) {
     public:
         int foo() const override { PYBIND11_OVERRIDE(int, ProtectedB, foo, ); }
         void *void_foo() override { PYBIND11_OVERRIDE(void *, ProtectedB, void_foo, ); }
+        void *get_self() override { PYBIND11_OVERRIDE(void *, ProtectedB, get_self, ); }
     };
 
     class PublicistB : public ProtectedB {
@@ -404,6 +406,7 @@ TEST_SUBMODULE(class_, m) {
         ~PublicistB() override{}; // NOLINT(modernize-use-equals-default)
         using ProtectedB::foo;
         using ProtectedB::void_foo;
+        using ProtectedB::get_self;
     };
 
     m.def("read_foo", [](const void *original) {
@@ -411,10 +414,15 @@ TEST_SUBMODULE(class_, m) {
         return *ptr;
     });
 
+    m.def("pointers_equal", [](const void *original, const void* comparison) {
+        return original == comparison;
+    });
+
     py::class_<ProtectedB, TrampolineB>(m, "ProtectedB")
         .def(py::init<>())
         .def("foo", &PublicistB::foo)
-        .def("void_foo", &PublicistB::void_foo);
+        .def("void_foo", &PublicistB::void_foo)
+        .def("get_self", &PublicistB::get_self);
 
     // test_brace_initialization
     struct BraceInitialization {

--- a/tests/test_class.cpp
+++ b/tests/test_class.cpp
@@ -384,8 +384,8 @@ TEST_SUBMODULE(class_, m) {
 
     protected:
         virtual int foo() const { return value; }
-        virtual void *void_foo() { return (void *) &value; }
-        virtual void *get_self() { return (void *) this; }
+        virtual void *void_foo() { return static_cast<void *>(&value); }
+        virtual void *get_self() { return static_cast<void *>(this); }
 
     private:
         int value = 42;

--- a/tests/test_class.cpp
+++ b/tests/test_class.cpp
@@ -384,6 +384,7 @@ TEST_SUBMODULE(class_, m) {
 
     protected:
         virtual int foo() const { return value; }
+        virtual void *void_foo() { return (void *) &value; }
 
     private:
         int value = 42;
@@ -392,6 +393,7 @@ TEST_SUBMODULE(class_, m) {
     class TrampolineB : public ProtectedB {
     public:
         int foo() const override { PYBIND11_OVERRIDE(int, ProtectedB, foo, ); }
+        void *void_foo() override { PYBIND11_OVERRIDE(void *, ProtectedB, void_foo, ); }
     };
 
     class PublicistB : public ProtectedB {
@@ -401,11 +403,18 @@ TEST_SUBMODULE(class_, m) {
         // (in Debug builds only, tested with icpc (ICC) 2021.1 Beta 20200827)
         ~PublicistB() override{}; // NOLINT(modernize-use-equals-default)
         using ProtectedB::foo;
+        using ProtectedB::void_foo;
     };
+
+    m.def("read_foo", [](const void *original) {
+        const int *ptr = reinterpret_cast<const int *>(original);
+        return *ptr;
+    });
 
     py::class_<ProtectedB, TrampolineB>(m, "ProtectedB")
         .def(py::init<>())
-        .def("foo", &PublicistB::foo);
+        .def("foo", &PublicistB::foo)
+        .def("void_foo", &PublicistB::void_foo);
 
     // test_brace_initialization
     struct BraceInitialization {

--- a/tests/test_class.cpp
+++ b/tests/test_class.cpp
@@ -405,8 +405,8 @@ TEST_SUBMODULE(class_, m) {
         // (in Debug builds only, tested with icpc (ICC) 2021.1 Beta 20200827)
         ~PublicistB() override{}; // NOLINT(modernize-use-equals-default)
         using ProtectedB::foo;
-        using ProtectedB::void_foo;
         using ProtectedB::get_self;
+        using ProtectedB::void_foo;
     };
 
     m.def("read_foo", [](const void *original) {
@@ -414,9 +414,8 @@ TEST_SUBMODULE(class_, m) {
         return *ptr;
     });
 
-    m.def("pointers_equal", [](const void *original, const void* comparison) {
-        return original == comparison;
-    });
+    m.def("pointers_equal",
+          [](const void *original, const void *comparison) { return original == comparison; });
 
     py::class_<ProtectedB, TrampolineB>(m, "ProtectedB")
         .def(py::init<>())

--- a/tests/test_class.py
+++ b/tests/test_class.py
@@ -1,5 +1,3 @@
-import ctypes
-
 import pytest
 
 import env  # noqa: F401
@@ -316,6 +314,7 @@ def test_bind_protected_functions():
     b = m.ProtectedB()
     assert b.foo() == 42
     assert m.read_foo(b.void_foo()) == 42
+    assert m.pointers_equal(b.get_self(), b)
 
     class C(m.ProtectedB):
         def __init__(self):

--- a/tests/test_class.py
+++ b/tests/test_class.py
@@ -1,3 +1,5 @@
+import ctypes
+
 import pytest
 
 import env  # noqa: F401
@@ -313,6 +315,7 @@ def test_bind_protected_functions():
 
     b = m.ProtectedB()
     assert b.foo() == 42
+    assert m.read_foo(b.void_foo()) == 42
 
     class C(m.ProtectedB):
         def __init__(self):

